### ISSUE TITLE
Big performance regression in Renovate 28.11.1 because of `re2`

### ISSUE THREAD

**Issue by onigoetz** [renovatebot/renovate#12736]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

28.11.1

### Please select which platform you are using if self-hosting.

GitHub Enterprise Server

### If you're self-hosting Renovate, tell us what version of the platform you run.

GHE 3.2.1

### Describe the bug

As mentioned in : https://github.com/renovatebot/renovate/discussions/12722

It seems that the PR to move many RegExes to `re2` introduced a significant slowdown on Renovate
I have an install of Renovate running for around 600 repositories, scheduled to run every hour.
With version 28.11.0 It was running in 40 minutes
With Renovate 28.11.1 It took 1 hour to do around 200 repositories, so around 4.5 times slower.

All versions after that (I tested up to 29.8.3) have the same issue.

### The fix I applied

To confirm this I did the following : 

Add this to the `Dockerfile`

```
# Patch the regex lib
RUN mv node_modules/renovate/dist/util/regex.js node_modules/renovate/dist/util/regex-original.js
COPY overrides/util_regex.js node_modules/renovate/dist/util/regex.js
```

`overrides/util_regex.js`:

```
const originalRegex = require("./regex-original.js");

const cache = new Map();

function regEx(pattern, flags) {
  const key = `${pattern.toString()}:${flags}`;

  if (cache.has(key)) {
    return cache.get(key);
  }

  const instance = originalRegex.regEx(pattern, flags);

  cache.set(key, instance);

  return instance;
}

module.exports = {
  ...originalRegex,
  regEx
};
```

This change is barbaric, I'll give you that, but it allows to reuse RegEx instances instead of re-creating them on every call.
With this change applied to both Renovate 28.11.1 and 29.8.3 I was able to run all my 600 repositories in 40 minutes again.

A more sensible fix would be to move all the `regEx()` function calls to the module level in order to do them only once.

### Relevant debug logs

The relevant logs are in the bug description above

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by rarkins** [comment]
I think that's a clever workaround and worth applying immediately. @viceice WDYT?

**Comment by viceice** [comment]
this can cause side effects on `string.replace` as the passed regex can't be reused. Not sure if this still applies to modern node v14+

**Comment by onigoetz** [comment]
What do you mean by string.replace and RegEx reuse ? it just answers with a RegEx instance, it still allows to use that instance more than once.

**Comment by rarkins** [comment]
PR: #12738

I ran it on a small repo and added some temporary stats and found it had a cache hit ratio of about 70%. For bigger repos with more dependencies I assume it's even higher.

**Comment by viceice** [comment]
https://stackoverflow.com/questions/2141974/javascript-regex-literal-with-g-used-multiple-times

**Comment by rarkins** [comment]
If we think it can be a concern, then we could exclude global patterns from the cache, however there's at least two ways to define tags.

**Comment by onigoetz** [comment]
> https://stackoverflow.com/questions/2141974/javascript-regex-literal-with-g-used-multiple-times

Oh thanks for the link, I didn't know about that.

BTW maybe one remark, is it important to reset the cache between runs ? Just to be sure that the cache doesn't grow infinitely

**Comment by viceice** [comment]
One more interesting thing, regex like `/(b)/g` are statically compiled, so the same string is reused all the time, so it's all the same regex instance.

So caching should be ok as fast perf fix.

**Comment by viceice** [comment]
![image](https://user-images.githubusercontent.com/1798109/142433916-bcc6ca16-f361-4c53-a2fc-86661d657183.png)
So the issue from no longer seems to apply to nodejs

**Comment by rarkins** [comment]
> BTW maybe one remark, is it important to reset the cache between runs ? Just to be sure that the cache doesn't grow infinitely

I think it should stay the same size, shouldn't it?

**Comment by rarkins** [comment]
Let's close this once we confirm the new release is working and performance returned

**Comment by onigoetz** [comment]
> I think it should stay the same size, shouldn't it?

I don't know if new regexes are created on-the-fly with custom parameters (user provided patterns)

**Comment by rarkins** [comment]
Not seeing too much of a performance boost in the hosted app. Let me know what you find from your own testing.

**Comment by onigoetz** [comment]
Hmm that's curious. On my side most repos are 99% of maven (pom.xml) + yarn/npm packages.json files.
Maybe on that subset the performance difference is visible ?

**Comment by rarkins** [comment]
I think repos usually lean towards npm anyway, but maybe maven makes some difference

**Comment by onigoetz** [comment]
Sorry it didn't improve anything for you then :/

I'll deploy your latest release tomorrow morning and confirm if it fixes the issue on my side (waiting for the Docker image to be released)

**Comment by viceice** [comment]
Docker images for 29.13.4 released. 🤗

**Comment by viceice** [comment]
@rarkins it looks like the issue description mixed v28 and v29 versions. 😏

**Comment by onigoetz** [comment]
Confirmed that the issue is fixed with the docker image 29.13.3. 🤗  thanks for the quick fix !

> @viceice it looks like the issue description mixed v28 and v29 versions. 😏

No my issue was indeed with Renovate 28.11.1, it appeared when we upgraded from 27.21.0 to 29.3.3.
We directly noticed the performance drop. (Yeah next time we'll upgrade faster)
We first tried to upgrade to 29.8.3 see if it was an issue specific to 29.3.3 but that wasn't the case.
I then tried many other versions to track down where the issue came from exactly (27.31.4, 27.31.5, 28.0.1, 28.1.3, 28.10.0, 28.11.0, 28.11.1, 28.16.0, 28.17.0, 28.24.0, not in that order)

That's how I was sure the issue was introduced in 28.11.1, since there were three changes in that version, including two changes for go (we don't use go) it was obvious it was the regex thing.

That's when I tried to patch the version 28.11.1 with my workaround => it now worked perfectly
To make sure it was reliable I also applied the workaround to 29.8.3 => worked as well
